### DISCUSSION TITLE
I am having trouble using winget with packer.

### DISCUSSION THREAD

**Discussion by 5nines** [microsoft/winget-cli#3326] [Q&A]
I am trying to install the Adobe Reader package using a PowerShell provisioner within my packer config. In short, when I execute the winget install command from within the PowerShell provisioner, it always fails with the "0x8a15000f: Data required by the source is missing" message. I can pause the packer build, RDP into the build VM, execute the exact same command from a PowerShell prompt, and it works without issue. I have tried the latest winget version and a couple pre-release versions. I believe it is something related to PS remoting but just have not been able to find anything while scouring the web for answers. Maybe something that logging into the machine interactively fixes. Any help would be greatly appreciated.

winget install --id Adobe.Acrobat.Reader.32-bit --accept-package-agreements --accept-source-agreements

**Comment by ghost** [comment]
Hi, were you able to get a solution? I have been trying to install WSL using winget through packer but packer doesnt even recognize winget install. what did you do to get winget install to work through packer?

**Comment by 5nines** [comment]
I have been able to install both the current and pre-release versions of winget through packer without any issues but am still unable to actually install any packages with the same issue described above. Hope the following may help you in your pursuit though:

**Current Version:**
`provisioner "powershell" {
  inline = [
    "Add-AppxPackage -RegisterByFamilyName -MainPackage Microsoft.DesktopAppInstaller_8wekyb3d8bbwe"
  ]
}`

**Pre-release version:**
`provisioner "powershell" {
  elevated_user = var.winrm_username
  elevated_password = build.WinRMPassword
  inline = [
    "$downloadURL = 'https://github.com/microsoft/winget-cli/releases/download/v1.6.1573-preview/Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle'",
    "$outFile = 'D:\\build\\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle'",
    "$wc = New-Object Net.WebClient",
    "$wc.DownloadFile($downloadURL, $outFile)",
    "Add-AppxPackage -Path 'D:\\Build\\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle'",
    "winget source remove --name msstore",
    "winget source update --name winget"
  ]
}`

I don't believe the elevated user is needed but I have tried both ways and it works. I included an example for each just in case. Build.WinRMPassword is a native packer variable for the auto-generated password local admin(packer) account.

**Reply by ghost** [reply]
Thank you for sharing. I actually figured out how to get winget installed but it took a longer way to do it. ill include that below: 

Write-Host 'Install NuGet'
"Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force"

Write-Host "Install UI.XAML"
$url = "https://www.nuget.org/api/v2/package/Microsoft.UI.Xaml/2.7.1"
$zipFile = "Microsoft.UI.Xaml.2.7.1.nupkg.zip"
Invoke-WebRequest -Uri $url -OutFile $zipFile
Expand-Archive $zipFile
Add-AppxPackage "Microsoft.UI.Xaml.2.7.1.nupkg\tools\AppX\x64\Release\Microsoft.UI.Xaml.2.7.appx"


Write-Host "Install VC.Libs"
Invoke-WebRequest -Uri https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx -OutFile Microsoft.VCLibs.x64.14.00.Desktop.appx
Add-AppxPackage 'https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx'


Write-Host "Install WinGet"
$latestWingetMsixBundleUri = $(Invoke-RestMethod https://api.github.com/repos/microsoft/winget-cli/releases/latest).assets.browser_download_url | Where-Object {$_.EndsWith(".msixbundle")}
$latestWingetMsixBundle = $latestWingetMsixBundleUri.Split("/")[-1]
Write-Information "Downloading winget to artifacts directory..."
Invoke-WebRequest -Uri $latestWingetMsixBundleUri -OutFile "./$latestWingetMsixBundle"
Add-AppxPackage $latestWingetMsixBundle

Now i am trying to install ubuntu, which works when i try it through my local machine but still gotta test through packer. 

I will keep you updated if i figure it out. 

Thank you for your help.

**Reply by 5nines** [reply]
Thank you for sharing as well. Have you been able to install any other packages with Winget through packer?

**Reply by ghost** [reply]
Nope not yet. im trying to install ubuntu. but need WSL for that which i installed through choco. but my pipeline keeps running into errors. so nothing so far.

**Reply by ghost** [reply]
When i run commands like these: 

 Write-Host "Installing Rancher"
 winget install rancher-desktop --silent --accept-source-agreements --accept-package-agreements

I keep getting this error: 

Installing Rancher
    azure-arm: The `msstore` source requires that you view the following agreements before using.
    azure-arm: Terms of Transaction: https://aka.ms/microsoft-store-terms-of-transaction
    azure-arm: The source requires the current machine's 2-letter geographic region to be sent to the backend service to function properly (ex. "US").
    azure-arm:
    azure-arm: Failed when searching source: winget
    azure-arm: An unexpected error occurred while executing the command:
    azure-arm: 0x8a15000f : Data required by the source is missing
    azure-arm:
    azure-arm: No packages were found among the working sources.
==> azure-arm: Provisioning step had errors: Running the cleanup provisioner, if present...

I just saw your original post and ur running into the same thing cause we using the same commands

**Comment by 5nines** [comment]
That is the exact error and experience I have.

________________________________

**Reply by ghost** [reply]
hey i figured out how to get past that error. but it then gives me this error: 0x80070520 : A specified logon session does not exist. It may already have been terminated. 

just add this to the end of your install:     --source msstore

**Comment by benlye** [comment]
Seems to be a known issue:
https://github.com/microsoft/winget-cli/issues/256

**Comment by florian-de** [comment]
Is there anyone who has had success installing winget applications in a packer build on windows 11 23h2?

**Reply by Lucero7919** [reply]
Yes, what help do you need?

**Reply by denelon** [reply]
Hey, I'd be curious to know if you're using the WinGet CLI or the PowerShell cmdlets for this scenario. It might make sense to document the steps to make it easy for others to follow.

**Reply by Lucero7919** [reply]
In essence, there are a few steps. During OOBE, Intune, SCCM, Packer, whatever, unless winget is running under the SYSTEM context, it's never going to work. However, winget was never really designed to be used in this manner. Arguable, chocolately is better. However, people prefer "native" so want to use winget. 

The first challenge: Setting the script to the DesktopApp Installer. If this is a fresh build and you type "winget" at a normal prompt, you'll get an error (path, cmdlet, whatever) - it's not going to find it. However, if this was a normal build of Win10/11 and you logged on to the desktop and had an active connection, what actually happens is DesktopAppInstaller gets patched in the background, a new version is added and the term 'winget' is added to the Path. You don't get this in an OSD, Packer, OOBE mode. 

So, in these modes we have to download it update first, then apply it, then set the location to the DesktopAppInstaller path, then you can use winget as intended, under SYSTEM context. I am in the process of enhancing this further, but for now here is a noddy script that will get you where you want to be:

```
#WebClient
$dc = New-Object net.webclient
$dc.UseDefaultCredentials = $true
$dc.Headers.Add("user-agent", "Inter Explorer")
$dc.Headers.Add("X-FORMS_BASED_AUTH_ACCEPTED", "f")

#temp folder
$InstallerFolder = $(Join-Path $env:ProgramData CustomScripts)
if (!(Test-Path $InstallerFolder)) {
    New-Item -Path $InstallerFolder -ItemType Directory -Force -Confirm:$false
}
#Check Winget Install
Write-Host "Checking if Winget is installed" -ForegroundColor Yellow
$TestWinget = Get-AppxProvisionedPackage -Online | Where-Object { $_.DisplayName -eq "Microsoft.DesktopAppInstaller" }
If ([Version]$TestWinGet. Version -gt "2022.506.16.0") {
    Write-Host "WinGet is Installed" -ForegroundColor Green
}
Else {
    #Download WinGet MSIXBundle
    Write-Host "Not installed. Downloading WinGet..." 
    $WinGetURL = "https://aka.ms/getwinget"
    $dc.DownloadFile($WinGetURL, "$InstallerFolder\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle")
		
    #Install WinGet MSIXBundle 
    Try {
        Write-Host "Installing MSIXBundle for App Installer..." 
        Add-AppxProvisionedPackage -Online -PackagePath "$InstallerFolder\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle" -SkipLicense 
        Write-Host "Installed MSIXBundle for App Installer" -ForegroundColor Green
    }
    Catch {
        Write-Host "Failed to install MSIXBundle for App Installer..." -ForegroundColor Red
    } 
	
    #Remove WinGet MSIXBundle 
    Remove-Item $InstallerFolder -Force -Recurse 
}

################################

$ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
if ($ResolveWingetPath) {
    $WingetPath = $ResolveWingetPath[-1].Path
}

Set-Location $wingetpath

$wingetApps = @(
    @{ ID = "Microsoft.VCRedist.2015+.x86"; MachineScope = $true },
    @{ ID = "Microsoft.VCRedist.2015+.x64"; MachineScope = $true },
    @{ ID = "Microsoft.VisualStudioCode"; MachineScope = $true }
)

$wingetpath = $wingetpath + "\winget.exe"

ForEach ($app in $wingetApps) {

    $argument = "install --exact --id $($app.ID) --silent --accept-package-agreements --accept-source-agreements"

    if ($app.MachineScope) {
        $argument = $argument + " --scope machine"
    }

    if ($app.Arguments) {
        $argument = $argument + " --override `"$($app.Arguments)`""
    }

    if ($app.ForceRequired) {
        $argument = $argument + " --force"
    }

    if ($app.Version) {
        $argument = $argument + " --version $($app.Version)"
    }

    $proc = start-process $wingetpath -ArgumentList $argument -PassThru
    $proc.WaitForExit();

    if ($proc.ExitCode -ne 0) {
        Write-Warning "$($app.ID) exited with code $($proc.ExitCode)"
    }
    else {  
        Write-Host "Install of $($app.ID) completed successfully"
    }

}
.\winget.exe list --source winget

```

Further challenges, for some deployment methods, wider pre-reqs are required including:
MS Visual C++ Redistributable
Microsoft.VCLibs.140.00.UWPDesktop
Microsoft.UI.Xaml.2.8

So if you want a smoother install, you'd need to include something like this for each of the pre-reqs listed:

```
    if (!($path)) {
        Write-Host "Installing MS Visual C++ 2015-2022..." -ForegroundColor Red
        $OSArch = if ($env:PROCESSOR_ARCHITECTURE -like "*64*") { "x64" } else { "x86" }
        $SourceURL = "https://aka.ms/vs/17/release/VC_redist.$OSArch.exe"
        $Installer = "$env:TEMP\VC_redist.$OSArch.exe"
        Invoke-WebRequest $SourceURL -OutFile $Installer -UseBasicParsing
        Start-Process -FilePath $Installer -Args "/quiet /norestart" -Wait
        Remove-Item $Installer -ErrorAction Ignore
    }
```

**Reply by denelon** [reply]
Interesting! Thanks for sharing.

The Repair-WinGetPackageManager cmdlet was designed to help "bootstrap" WinGet or get it to the latest version. That "should" work in system context. Once that's done, the PowerShell cmdlets in the Microsoft.WinGet.Client PowerShell module should work. Of course there are challenges when in system context related to only being able to install things that are "machine wide" since there is no user, and portable packages require modifying the settings to use tar.exe rather than the Windows shell implementation of zip.

**Reply by mloskot** [reply]
For the records, I've just tried to build Windows 11 image using Packer and Azure Image Builder, and to enable WinGet with `Microsoft.WinGet.Client` module, and this snippet, after [the docs](https://learn.microsoft.com/en-us/windows/package-manager/winget/#install-winget-on-windows-sandbox):

```powershell
Install-PackageProvider -Name NuGet -Force | Out-Null
Install-Module -Name Microsoft.WinGet.Client -Force -Repository PSGallery | Out-Null
Repair-WinGetPackageManager -AllUsers
```

and it failed:

```
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11: Repair-WinGetPackageManager : One or more errors occurred.
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11: At C:\Windows\Temp\script-6973eb1b-ba40-0658-f9a5-1d00b1289424.ps1:8 char:1
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11: + Repair-WinGetPackageManager -AllUsers
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11:     + CategoryInfo          : NotSpecified: (:) [Repair-WinGetPackageManager], AggregateException
│2026-01-23T22:43:43+01:00: ==> azure-arm.win11:     + FullyQualifiedErrorId : System.AggregateException,Microsoft.WinGet.Client.Commands.RepairWinGetPackageManagerCmdlet
```
